transhumandesign/kag-base
# Fixed bugs and improved consistency of Trader, Migrant and Princess

## Status

- **READY**: this PR is (to the best of your knowledge) ready to be incorporated into the game.

## Listing

[fixed] Trader corpse flying through the sky when killed on a ladder.
[fixed] Characters not receiving damage from explosions in their inventory.
[fixed] Trader corpse gliding over the ground when thrown
[fixed] Trader corpse didn't decay in storages/crates/etc.
[fixed] Trader corpse making a grunt sound when hitting the ground hard.
[fixed] Trader corpse not showing the correct icon when getting loaded on a Catapult.
[removed] Being able to put Trader corpse in your backpack
[changed] Migrant has their own death scream (MigrantScream1 instead of WilhelmShort).
[changed] Princess corpse can be put in inventories

## Description

This PR aims to fix some bugs and improve consistency of Migrant, Trader and Princess.
If I didn't know better, Trader may be a relict of the old days. But it is fun to play with in Sandbox and so I decided to fix them.

Characters that didn't have the "player" tag didn't get hurt by exploding bombs in their inventory.
I changed `hasTag("player")` to `hasTag("flesh")` in Explosion.as. 
Although there may be a better way, I figured Bison, Shark, Migrant, Chicken, Fishy, and whatever else can have the "flesh" tag will never have bombs explode in their inventory. So there shouldn't be a side-effect.

RunnerDeath.as has been added to Trader.cfg scripts. This allows the following fixes:
- Cannot look in Trader's backpack anymore.
- Trader corpse doesn't glide over the ground when thrown.
- Trader corpse cannot be put in player's inventory.
- Trader corpse expires after 45 sec (set by RunnerCommon.as) instead of gibbing after 20 sec (set by Trader.as).
- Trader corpse doesn't fly through the sky if killed on a ladder.
- Trader expires in inventory, instead of staying.
- Added hasTag("no death sound") check in RunnerDeath.as.
It is used by Trader and Migrant so they can play their own death cries without the default WilhemShort.ogg getting played.

Removed onHealthChange() from Trader.as, let RunnerDeath.as handle damages and death.
- This causes Trader to be dead at 0 health, instead of at < 1.0 health.
- Therefore, there is no cry sound when the corpse falls on the ground hard while it is above 0 health. 

Princess corpse can now be put in storages, but not in the player's inventory.

Loading Trader's corpse on a Catapult shows the correct icon.
This was done by commenting out an AddIconToken() instance in TradingMenu.as.


## Steps to Test or Reproduce

1. Go to Sandbox.
2. Spawn a Trader and kill him.
3. Throw him.
4. Notice he will glide over the ground. **After this change, he will behave like any other corpse**

----

1. Go to Sandbox.
2. Build a ladder and climb on it.
3. Spawn a Trader and kill him while the ladder stays intact.
4. Notice the Trader will fly through the sky. **After this change, he will fall down instead**

----

1. Go to Sandbox.
2. Spawn a Trader.
3. Show buttons. Notice you can look in the Trader's backpack. **After this change, you can't**

----

1. Go to Sandbox.
2. Spawn a Trader.
3. Kill him.
4. Notice he can be put in your inventory. **After this change, you can't**

----

1. Go to Sandbox.
2. Spawn a Trader.
3. Kill him.
4. Put him in a Crate.
5. Check back in a minute.
6. Notice he will still be in there. **After this change, he will be gone after 45 sec**

----

1. Go to Sandbox.
2. Spawn a Trader.
3. Kill him so he has 0.75 health left.
4. Throw him on the ground from some height, so he lands hard.
5. Notice he will shout. **After this change, the corpse doesn't shout.**

----

1. Go to Sandbox.
2. Spawn a Princess Geti.
3. Kill him.
4. Notice you can't put him in Storages/Crates/etc. **After this change, you can.**

----

1. Go to Sandbox.
2. Spawn a Trader.
3. Kill him.
4. Load him on a Catapult.
5. Notice he icon will be a small Trading Post. **After this change, it will be a Trader icon.**
